instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/asyncannotationsupport/AsyncOperationEndSupport.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.asyncannotationsupport;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.Instrumenter;
+
+/**
+ * A wrapper over {@link Instrumenter} that is able to defer {@link Instrumenter#end(Context,
+ * Object, Object, Throwable)} until asynchronous computation finishes.
+ *
+ * <p>This class is not able to extract {@code RESPONSE} from the asynchronous computation value it
```

Comment:
I don't quite follow - why is this? Is it because some return types don't have one response (e.g., a reactive stream that isn't a mono)? Can we at least resolve it for the mono-like cases (especially CompletableFuture)?

Reply:
> Is it because some return types don't have one response (e.g., a reactive stream that isn't a mono)?

That was one reason; the other "reason" (well, "excuse" would be more correct 😅 ) was that all those strategies pretty much operate on raw/unchecked types, before passing the response to the Instrumenter I'd have to check if it's an instance of RESPONSE type - I'd need to pass `Class<RESPONSE>` somewhere to do that, and I wanted to keep those strategies as simple as possible.
I'll work on that; multi-result reactive types should be the only thing that's not passed to the instrumenter.

Reply:
Having gone down the Java generics and reflection rabbit hole recently this sounds like it would be very tricky to really get right. Would the strategy itself be responsible for trying to negotiate the actual type parameters of the returned generic type?

Reply:
I thought to make it as simple as possible, just use a `Class<RESPONSE>` instance to check type and cast, without delving into Java generics weirdness. All `Instrumenter`s in this project that would use the async support would use `Void` as the response type, so we'd never actually need to pass anything there - at least in this repo.

Reply:
Maybe I'm not understanding.  Is the goal to be able to derive that `RESPONSE` is `String` and to pass a `Class<String>` to the instrumentation if the annotated method returns `CompletableFuture<String>` or `Mono<String>` or the like?  Or is this for non-AOP instrumentation?

Reply:
I went with the simplest possible approach:

```java
  public static <REQUEST, RESPONSE> AsyncOperationEndSupport<REQUEST, RESPONSE> create(
      Instrumenter<REQUEST, RESPONSE> syncInstrumenter,
      Class<RESPONSE> responseType,
      Class<?> asyncType)
```

You have to manually pass the expected response type. Extracting it from the passed `Instrumenter` is impossible, it will never have type parameters in the runtime.